src/plugins.c

```diff
@@ -387,68 +530,67 @@ plugin_new(const gchar *fname, gboolean load_plugin, gboolean add_to_list)
 		return NULL;
 	}
 
-	if (! plugin_check_version(module))
+	plugin = g_new0(Plugin, 1);
+	plugin->module = module;
+	plugin->filename = g_strdup(fname);
+	plugin->public.geany_data = &geany_data;
+	plugin->public.priv = plugin;
+	/* Fields of plugin->info/hooks must to be initialized by the plugin */
+	plugin->public.info = &plugin->info;
+	plugin->public.hooks = &plugin->hooks;
+
+	g_module_symbol(module, "geany_load_module", (void *) &p_geany_load_module);
+	if (p_geany_load_module)
 	{
-		if (! g_module_close(module))
-			g_warning("%s: %s", fname, g_module_error());
-		return NULL;
+		/* This is a new style plugin. It should fill in plugin->info and then call
+		 * geany_plugin_register() in its geany_load_module() to successfully load.
+		 * The ABI and API checks are performed by geany_plugin_register() (i.e. by us).
+		 * We check the LOADED_OK flag separately to protect us against buggy plugins
+		 * who ignore the result of geany_plugin_register() and register anyway */
+		p_geany_load_module(&plugin->public, module, GEANY_API_VERSION);
 	}
-
-	g_module_symbol(module, "plugin_set_info", (void *) &plugin_set_info);
-	if (plugin_set_info == NULL)
+	else
 	{
-		geany_debug("No plugin_set_info() defined for \"%s\" - ignoring plugin!", fname);
-
-		if (! g_module_close(module))
-			g_warning("%s: %s", fname, g_module_error());
-		return NULL;
+		/* This is the legacy / deprecated code path. It does roughly the same as
+		 * geany_load_module() and geany_plugin_register() together for the new ones */
+		register_legacy_plugin(plugin, module);
 	}
 
-	plugin = g_new0(Plugin, 1);
-
-	/* set basic fields here to allow plugins to call Geany functions in set_info() */
-	g_module_symbol(module, "geany_data", (void *) &p_geany_data);
-	if (p_geany_data)
-		*p_geany_data = &geany_data;
-
-	/* read plugin name, etc. */
-	plugin_set_info(&plugin->info);
-	if (G_UNLIKELY(EMPTY(plugin->info.name)))
+	if (! PLUGIN_LOADED_OK(plugin))
 	{
-		geany_debug("No plugin name set in plugin_set_info() for \"%s\" - ignoring plugin!",
-			fname);
-
-		if (! g_module_close(module))
-			g_warning("%s: %s", fname, g_module_error());
-		g_free(plugin);
-		return NULL;
+		geany_debug("Failed to load \"%s\" - ignoring plugin!", fname);
+		goto err;
 	}
 
-	g_module_symbol(module, "plugin_init", (void *) &plugin->init);
-	if (plugin->init == NULL)
+	if (EMPTY(plugin->info.name))
 	{
-		geany_debug("Plugin '%s' has no plugin_init() function - ignoring plugin!",
-			plugin->info.name);
-
-		if (! g_module_close(module))
-			g_warning("%s: %s", fname, g_module_error());
-		g_free(plugin);
-		return NULL;
+		geany_debug("No plugin name set for \"%s\" - ignoring plugin!", fname);
+		goto err;
 	}
-	/*geany_debug("Initializing plugin '%s'", plugin->info.name);*/
 
```

Comment:
if you did remove this empty line, the commented out `geany_debug()` wouldn't be changed in the diff :] (who cares though)